ipa-core/src/protocol/context/dzkp_malicious.rs

```diff
@@ -0,0 +1,151 @@
+use std::{
+    fmt::{Debug, Formatter},
+    num::NonZeroUsize,
+};
+
+use async_trait::async_trait;
+
+use crate::{
+    error::Error,
+    helpers::{ChannelId, Gateway, Message, ReceivingEnd, Role, SendingEnd, TotalRecords},
+    protocol::{
+        context::{
+            dzkp_validator::DZKPBatch, prss::InstrumentedIndexedSharedRandomness, Base,
+            Context as ContextTrait, DZKPContext, InstrumentedSequentialSharedRandomness,
+        },
+        prss::Endpoint as PrssEndpoint,
+        step::{Gate, Step, StepNarrow},
+    },
+    seq_join::SeqJoin,
+    sync::Arc,
+};
+
+/// Represents protocol context in malicious setting, i.e. secure against one active adversary
+/// in 3 party MPC ring.
+#[derive(Clone)]
+pub struct DZKPUpgraded<'a> {
+    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
+    /// may operate with raw references and be more efficient
+    inner: Arc<DZKPUpgradedInner<'a>>,
+    gate: Gate,
+    total_records: TotalRecords,
+}
+
+impl<'a> DZKPUpgraded<'a> {
+    pub(super) fn new<S: Step + ?Sized>(
+        source: &Base<'a>,
+        malicious_step: &S,
+        batch: DZKPBatch,
+    ) -> Self
+    where
+        Gate: StepNarrow<S>,
+    {
+        Self {
+            inner: DZKPUpgradedInner::new(source, batch),
+            gate: source.gate().narrow(malicious_step),
+            total_records: TotalRecords::Unspecified,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a> DZKPContext for DZKPUpgraded<'a> {
+    fn is_safe(&self) -> Result<(), Error> {
```

Comment:
Can you add a comment explaining what it means to be "safe"?

Reply:
I added it to the trait definition. It just checks that there are currently no values that still need to be verified. So you could e.g. reveal shares or drop the validator